core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```
@@ -47,43 +47,48 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
     with WorkspaceSupport
     with LazyLogging {
 
-  def createSnapshot(workspaceName: WorkspaceName, snapshot: NamedDataRepoSnapshot): Future[DataRepoSnapshotResource] =
+  def createSnapshot(workspaceName: WorkspaceName,
+                     snapshotIdentifiers: NamedDataRepoSnapshot
+  ): Future[DataRepoSnapshotResource] =
     getV2WorkspaceContextAndPermissions(workspaceName,
                                         SamWorkspaceActions.write,
                                         Some(WorkspaceAttributeSpecs(all = false))
-    ).flatMap { workspaceContext =>
-      val wsid = workspaceContext.workspaceIdAsUUID // to avoid UUID parsing multiple times
-      // create the stub workspace in WSM if it does not already exist
-      if (!workspaceStubExists(wsid, ctx)) {
-        workspaceManagerDAO.createWorkspace(wsid, ctx)
+    ).flatMap { rawlsWorkspace =>
+      val wsid = rawlsWorkspace.workspaceIdAsUUID // to avoid UUID parsing multiple times
+      val snapshot = dataRepoDAO.getSnapshot(snapshotIdentifiers.snapshotId, ctx.userInfo.accessToken)
+      val validator = new SnapshotReferenceCreationValidator(rawlsWorkspace, snapshot)
+      val aggregatedWorkspaceService = new AggregatedWorkspaceService(workspaceManagerDAO)
```

hm - TODO here is to check with the workspaces team. The `Workspace` object has a `workspaceType: WorkspaceType` which I think is a valid proxy for Gcp vs. Azure. You may also be able to determine the cloud platform from the `googleProjectNumber` or `bucketName` fields. If Workspaces team says these are safe to rely on, you could eliminate the need for `AggregatedWorkspaceService` entirely, including the extra outbound REST call.

---

`WorkspaceType` looks like it's either `McWorkspace` (which, regretfully, I can no longer read as anything but MickWorkspace, sorry if I've thus infected you), or `RawlsWorkspace`.  I interpreted the `McWorkspace` type as potentially being GCP OR Azure (or in the futuretimes, both).

@aherbst-broad do you have any advice here?  I thought using `AggregatedWorkspaceService` might be the way to get from WSM a determination of whether a workspace is for Azure vs. GCP, but maybe I can shortcut it by just looking at `WorkspaceType` and obviate the need for `AggregatedWorkspaceService` entirely?

---

It'd preferable to leverage the AggregatedWorkspaceService here and rely on the truth from WSM if present. That is, the AggregatedWorkspaceService should Do The Right Thing™ (or be made to do so) and give you back a workspace with maybe the needed cloud platform information from WSM. If there is WSM info present, use that for decision making, if not, assume GCP. This would keep us from having to migrate one more thing as we move toward a WSM-based future.

---

good call on using `AggregatedWorkspaceService` here to future-proof the code - that makes sense. 

@jladieu what I'm really trying to avoid here is that if the current workspace is an Azure workspace, we're going to add a call to WSM and then immediately fail the import because we don't support Azure. What do you think about refactoring the code so that if the _snapshot_ being imported is an Azure snapshot, we fail fast before reaching out to WSM to determine the workspace's cloud?

---

Ok I totally understand your concern about adding an unnecessary call that we could avoid by failing early; I'll need to re-organize the flow a bit here but your comment below provides a good roadmap!  Thanks for digging into the detail on this.

---

Reorganized this quite a bit after incorporating input from TDR & Workspaces.  We should now fail fast on unsupported platforms without incurring a hit to WSM, and `AggregatedWorkspaceService` is able to support GCP Workspaces (even Rawlsy ones!) as a result of https://github.com/broadinstitute/rawls/pull/2563